crates/net/eth-wire/src/types/message.rs

```diff
@@ -237,6 +237,159 @@ impl EthMessage {
+    /// * `limit` - The maximum allowed size of the encoded data.
+    ///
+    /// # Returns
+    ///
+    /// A vector containing the encoded data, truncated to fit within the size limit if necessary.
+    /// Users of this method should be aware that the truncated data may not represent the complete
+    /// original dataset.
+    fn encode_truncate(&self, approx: usize, limit: usize) -> Vec<u8>;
+
+    /// Encodes the data up to the specified size limit, returning the encoded data along with
+    /// a flag indicating whether the limit was exceeded.
+    ///
+    /// # Arguments
+    ///
+    /// * `approx` - The approximate size to allocate in memory for the encoding process. This is
+    ///   used to pre-allocate memory for efficiency.
+    /// * `limit` - The maximum allowed size of the encoded data.
+    ///
+    /// # Returns
+    ///
+    /// A tuple containing:
+    /// * The encoded data as `Vec<u8>`.
+    /// * The length of the data at the point just before reaching the limit, or 0 if the limit was
+    ///   not exceeded.
+    /// * Actual payload length
+    fn encode_until_limit(&self, approx: usize, limit: usize) -> (Vec<u8>, usize, usize);
+}
+
+impl<T: Encodable> EncodableExt for Vec<T> {
+    fn encode_max(
+        &self,
+        approx: usize,
+        limit: usize,
+    ) -> alloy_rlp::Result<Vec<u8>, alloy_rlp::Error> {
+        let (mut buf, current_len, actual_payload_length) = self.encode_until_limit(approx, limit);
+
+        if current_len != 0 {
+            // Don't shrink if limit_exceeded since we return Error anyway
+            return Err(alloy_rlp::Error::Custom("Size limit exceeded"));
+        }
+        let mut header_buffer = Vec::new();
+        let header = Header { list: true, payload_length: actual_payload_length };
+        header.encode(&mut header_buffer);
+        // Ensure the encoded header does not exceed the placeholder size
+        assert!(header_buffer.len() <= 3); // header placeholder size is 3
+                                           // Replace the placeholder with the real header
+        buf[..header_buffer.len()].copy_from_slice(&header_buffer[..]);
+        // If the real header is smaller than the placeholder, remove the excess
+        if header_buffer.len() < 3 {
+            let excess = 3 - header_buffer.len();
+            buf.drain(header_buffer.len()..excess);
+        }
+        buf.shrink_to_fit();
+        Ok(buf)
+    }
+
+    fn encode_until_limit(&self, approx: usize, limit: usize) -> (Vec<u8>, usize, usize) {
+        let mut buffer = Vec::with_capacity(approx);
+        let mut header = Header { list: true, payload_length: 0 };
+        for item in self {
+            // initial payload length calculation if limit is not exceeded
+            header.payload_length += item.length();
+        }
+        // header.encode(&mut buffer);
+        let header_placeholder_size = 3;
+        buffer.extend(vec![0; header_placeholder_size]);
```

Comment:
where's this coming from?

Reply:
Assumed max header length is <=3

Reply:
could also assume more, like 4 or 5 .

Reply:
the header also has a length method, it does require the payload length to be calculated properly, but in any case, we shouldn't need to extend anything at this point or include any placeholders

Reply:
yup. Made a push to remove placeholder